orttraining/orttraining/python/training/ort_triton/_sorted_graph.py

```diff
@@ -0,0 +1,210 @@
+
+        # Sort the initializers in reference order.
+        # We try to reuse Triton module for different ONNX models with same graph structure,
+        # even the node args names in the models are different.
+        # Sorting the initializers can help to generate same model key for different ONNX models.
+        initializers = {}
+        for initializer in self._graph.initializer:
+            initializers[initializer.name] = initializer
+        self._sorted_initializers: List[TensorInfo] = []
+        for node in self._sorted_nodes:
+            for input in node.input:
+                if input in initializers:
+                    self._sorted_initializers.append(initializers[input])
+                    initializers.pop(input)
+
+        # Split nodes to constant nodes and non-constant nodes.
+        self._const_nodes: List[NodeProto] = [node for node in self._sorted_nodes if node.op_type == "Constant"]
+        self._sorted_nodes: List[NodeProto] = [node for node in self._sorted_nodes if node.op_type != "Constant"]
+
+    def __str__(self):
+        """
+        Generate a unique key for the model based on the graph structure, ignoring the node args names.
+        We try to reuse Triton module for different ONNX models with same graph structure.
+        """
+        graph_inputs = []
+        name_map = {}
+        for idx, input in enumerate(self._graph.input):
+            shape_str = str(self._input_shapes[idx]).replace(" ", "")
+            graph_inputs.append(f"({str(input.type.tensor_type.elem_type)},{shape_str})")
+            name_map[input.name] = f"i{idx}"
+        graph_inputs_str = ",".join(graph_inputs)
+
+        constants = []
+        for idx, initializer in enumerate(self._sorted_initializers):
+            data_str = np.array2string(to_numpy_array(initializer), separator=",").replace("\n", "").replace(" ", "")
+            constants.append(f"({initializer.data_type},{data_str})")
+            name_map[initializer.name] = f"c{idx}"
+
+        for idx, node in enumerate(self._const_nodes):
+            value_attr = get_attribute(node, "value")
+            data_str = np.array2string(to_numpy_array(value_attr), separator=",").replace("\n", "").replace(" ", "")
+            constants.append(f"({value_attr.data_type},{data_str})")
+            name_map[node.output[0]] = f"c{idx + len(self._sorted_initializers)}"
+        constants_str = ",".join(constants)
+
+        for idx, output in enumerate(self._graph.output):
+            name_map[output.name] = f"o{idx}"
+
+        nodes = []
+        for node_idx, node in enumerate(self._sorted_nodes):
+            inputs = []
+            for input in node.input:
+                inputs.append(name_map[input] if input in name_map else input)
+            inputs_str = ",".join(inputs)
+            outputs = []
+            for idx, output in enumerate(node.output):
+                if output in name_map:
+                    outputs.append(name_map[output])
```

Comment:
Does this case happen or it is to handle optional outputs?

Reply:
name_map contains all processed names, including graph inputs and outputs. Here it goes through all nodes, if the output of the node is graph output, then it must be in the name_map already, otherwise, it's activation, need to create a new variable name for it, just like the code below to use prefix "t_".